Recorder/OutputType.cpp

```diff
@@ -463,169 +131,15 @@ int to_index(const OutputType config) {
-    if(OutputType::IF1 == L) return "IF";
-    if(OutputType::IF2 == L) return "IF";
-    if(OutputType::IF3 == L) return "IF";
-    if(OutputType::IF4 == L) return "IF";
-    if(OutputType::IF5 == L) return "IF";
-    if(OutputType::IF6 == L) return "IF";
-    if(OutputType::IM1 == L) return "IF";
-    if(OutputType::IM2 == L) return "IF";
-    if(OutputType::IM3 == L) return "IF";
-    if(OutputType::GDF == L) return "GDF";
-    if(OutputType::GDF1 == L) return "GDF";
-    if(OutputType::GDF2 == L) return "GDF";
-    if(OutputType::GDF3 == L) return "GDF";
-    if(OutputType::GDF4 == L) return "GDF";
-    if(OutputType::GDF5 == L) return "GDF";
-    if(OutputType::GDF6 == L) return "GDF";
-    if(OutputType::GDM1 == L) return "GDM";
-    if(OutputType::GDM2 == L) return "GDM";
-    if(OutputType::GDM3 == L) return "GDM";
-    if(OutputType::GIF == L) return "GIF";
-    if(OutputType::GIF1 == L) return "GIF";
-    if(OutputType::GIF2 == L) return "GIF";
-    if(OutputType::GIF3 == L) return "GIF";
-    if(OutputType::GIF4 == L) return "GIF";
-    if(OutputType::GIF5 == L) return "GIF";
-    if(OutputType::GIF6 == L) return "GIF";
-    if(OutputType::GIM1 == L) return "GIF";
-    if(OutputType::GIM2 == L) return "GIF";
-    if(OutputType::GIM3 == L) return "GIF";
-    if(OutputType::HYDRO == L) return "HYDRO";
-    if(OutputType::MISES == L) return "MISES";
-    if(OutputType::EEQ == L) return "EEQ";
-    if(OutputType::EEEQ == L) return "EEEQ";
-    if(OutputType::PEEQ == L) return "PEEQ";
-    if(OutputType::KE == L) return "KE";
-    if(OutputType::SE == L) return "SE";
-    if(OutputType::VE == L) return "VE";
-    if(OutputType::NVE == L) return "NVE";
-    if(OutputType::K == L) return "K";
-    if(OutputType::M == L) return "M";
-    if(OutputType::ED == L) return "ED";
-    if(OutputType::VD == L) return "VD";
-    if(OutputType::SD == L) return "SD";
-    if(OutputType::ES == L) return "ES";
-    if(OutputType::VS == L) return "VS";
-    if(OutputType::SS == L) return "SS";
-    if(OutputType::DAMAGE == L) return "DAMAGE";
-    if(OutputType::DT == L) return "DT";
-    if(OutputType::DC == L) return "DC";
-    if(OutputType::PP == L) return "PP";
-    if(OutputType::VF == L) return "VF";
-    if(OutputType::HIST == L) return "HIST";
-    if(OutputType::LITR == L) return "LITR";
-    if(OutputType::YF == L) return "YF";
-    if(OutputType::BEAME == L) return "BEAME";
-    if(OutputType::BEAMS == L) return "BEAMS";
-    if(OutputType::AMP == L) return "AMP";
-    if(OutputType::NL == L) return "NL";
-
-    return "NL";
+    std::string sInput = to_name(L);
+    std::string Out = std::regex_replace(sInput, std::regex(R"([0-9])"), "");
+    if (Out == "UR") return "U";
+    if (Out == "VR") return "V";
+    if (Out == "AR") return "A";
+    if (Out == "MMR") return "MM";
+    if (Out == "RM") return "RF";
+    if (Out == "IM") return "IF";
+    if (Out == "GDM") return "GDF";
+    if (Out == "GIM") return "GIF";
+    return Out.c_str();
```

Comment:
The static analysis basically says `Out.c_str()` is not valid, since `Out` will be destroyed once leaving this scope.

This method is mainly used in VTK related functions, see: https://github.com/TLCFEM/suanPan/blob/06ce2d368b053cdd71cc5a3dc9227c32d7a0f20b/Element/Visualisation/vtkParser.cpp#L255

The `SetActiveScalars` method only accepts `const char*`.

I think a better solution here is change the signature to return a `string` such that:


```cpp
std::string to_category(const OutputType L) {
    std::string result = std::regex_replace(to_name(L), std::regex(R"([0-9])"), "");
    switch (result) {
        case "UR": return "U";
        case "VR": return "V";
        case "AR": return "A";
        case "MMR": return "MM";
        case "RM": return "RF";
        case "IM": return "IF";
        case "GDM": return "GDF";
        case "GIM": return "GIF";
        default: return result;
    }
}
```

The usages in `vtkParser.cpp` can be changed to `to_category(config.type).c_str()`.

Reply:
I just didn't want to mess with the output type of the function. That is why I converted it to c-style string upon return.

Reply:
> I just didn't want to mess with the output type of the function. That is why I converted it to c-style string upon return.

Yeah, I know, but it won't work here.

Reply:
Ok. I will keep this in mind. Thank you so much.